### ISSUE TITLE
libreJS support

### ISSUE THREAD

**Issue by bill-auger** [thelounge/thelounge#2318]
we have noticed that this program does not work well with the current version of libreJS - would you be interested in patches to support libreJS users?

https://www.gnu.org/software/librejs/free-your-javascript.html

**Comment by bill-auger** [comment]
ping @David_Hedlund

**Comment by MaxLeiter** [comment]
I think as long as the patches don't remove/change functionality or modify licensing the patches will be accepted. Can you provide an example?

**Comment by astorije** [comment]
Hey @bill-auger, would you mind clarifying what kind of patches you had in mind?

**Comment by bill-auger** [comment]
ping @davidhedlund

**Comment by bill-auger** [comment]
these are just metadata comments or weblabels - they should not affect behavior in any way see the link in the OP for what that entails

**Comment by davidhedlund** [comment]
@bill-auger Thanks again.

**Comment by xPaw** [comment]
I don't see this working out in any reasonable manner. Blindly adding a 'free software' comment to a built bundle would be just misleading because we have countless dependencies (and sub-dependencies) built into two bundle files.

LibreJS on it's own doesn't seem like a good idea, and unsurprisingly hasn't gained any traction in over 4 years.

If we're not getting an acknowledgement that you're willing to work on a PR, I'm inclined to close this issue.

**Comment by bill-auger** [comment]
sure, feel free to close this - it is not more than a feature request at this point - i or someone will open a PR if they decide to follow through with it

let me say though the very language you used there is alarming in of itself

the idea is not add vague placating comments - it is to specify precisely the actual licenses that the code is distributed under - but even if that were how it works, how could adding a 'free software' comment be misleading? is this not free software?

software needs licenses in order to be distributed, and those who distribute it need to know what that license is (for all parts); or else they are literally not allowed to distribute it

finally, if it is indeed the case that the project has countless dependencies (and sub-dependencies), far from an excuse for not verifying licenses, that would be a urgent justification for verify them

that is literally saying you (the project lead?) do not actually know what this project's dependencies are - yet you are distributing them, presumably without knowing what licenses they are published under or if they have any license at ll - that kind of situation make systems developers shriek in horror - i hope you do you realize the implications of being so glib with piling on dependencies

the libreJS meta-data mainly exists to clarify licensing terms - whether the client tool is well-adopted or not is not nearly as important - knowing the licenses of the software you distribute is something you should want for yourself; and libreJS is a good way to codify that knowledge for the benefit of others - that is especially important when it is others that will be doing most f the re-distributing to clients

**Comment by astorije** [comment]
Hey @bill-auger, thanks a lot for your detailed answer! I know very little about LibreJS so I am going to merely answer on dependencies and licensing.

We do indeed [have quite a bunch of dependencies (and sub-dependencies)](https://github.com/thelounge/thelounge/blob/7591088982d0cf453bf447572ebbff8301b374ee/package.json#L41-L98). We tend to be careful of the dependencies we use, for example I go through commit diffs when Greenkeeper suggests a dependency update, we often consider several options before settling for a new dependency, we avoid those 1-liner dependencies the JS ecosystem has, etc.
That being said, it is true that licenses are easily overlooked when bringing a new dependency, and missing something is incredibly easy for sub-dependencies. It seems common to me that "if it's on GitHub, I can use it". I don't want to make any further generalities, but I am guilty of having jumped on new libraries without doing my homework correctly on more than one occasion.

So, licenses are hard. I can't speak for the team, but I would welcome any help in verifying dependencies, improving or fixing potential mistakes, or identifying our licensing shortcomings in a broader sense.

I was curious, so I ran `license-checker --summary`:

```
├─ MIT: 985
├─ ISC: 115
├─ BSD-3-Clause: 42
├─ MIT*: 19
├─ Apache-2.0: 18
├─ BSD-2-Clause: 15
├─ BSD*: 10
├─ CC-BY-3.0: 3
├─ Unlicense: 3
├─ CC0-1.0: 2
├─ (BSD-2-Clause OR MIT OR Apache-2.0): 2
├─ Apache License, Version 2.0: 2
├─ CC-BY-4.0: 2
├─ BSD: 2
├─ Public Domain: 1
├─ Custom: https://github.com/dominictarr/event-stream: 1
├─ (BSD-3-Clause OR GPL-2.0): 1
├─ (MIT AND Zlib): 1
├─ (WTFPL OR MIT): 1
├─ MIT,Apache2: 1
├─ (MIT OR Apache-2.0): 1
├─ (OFL-1.1 AND MIT): 1
├─ Custom: http://w3c.github.io/IntersectionObserver/: 1
├─ (MIT AND CC-BY-3.0): 1
├─ AFLv2.1,BSD: 1
├─ BSD-3-Clause OR MIT: 1
├─ (GPL-2.0 OR MIT): 1
└─ MPL-2.0: 1
```

It seems insane, but when I'm comparing to a brand new project I'm involved with at work, with little setup yet, outputs very similar numbers and rows even though the tech stack is very different.
I also get similar results with a couple other FOSS projects, so it's very likely we are not an isolated case, showing a larger problem.
Also in all cases, MIT was very largely dominating the list of licenses reported, with a ratio similar to ours.

In conclusion, I do believe in free software, and on The Lounge we aim to protect our users' privacy whenever we can, but I would welcome any help that makes things better and does not make unreasonable sacrifices. As I mentioned above, if you were to open a PR, that would help us getting an idea of what this means for us exactly. I can't promise anything will be accepted however.

Thanks again!

**Comment by davidhedlund** [comment]
@astorije is license-checker part of the package licenseutils? If not, where did you install it from?

**Comment by astorije** [comment]
@davidhedlund, I used https://www.npmjs.com/package/license-checker very quickly, but I don't doubt there are better tools out there. I just wanted a quick overview :)

**Comment by davidhedlund** [comment]
@astorije Thanks. I'll check it out.

You might want to compare it with [licenseutils](https://directory.fsf.org/wiki/Licenseutils) to see if it satisfies your needs.

**Comment by astorije** [comment]
Thanks!
Note however that, for a Node/npm project, the convenience of simply running `yarn add license-checker && yarn license-checker --summary` is hard to compete with :)
Feel free to post your own results and/or suggest improvements we could make, if you want to help us out!

**Comment by davidhedlund** [comment]
@astorije 

> Note however that, for a Node/npm project, the convenience of simply running yarn add license-checker && yarn license-checker --summary is hard to compete with :)

Ok, I understand.

**Comment by davidhedlund** [comment]
https://github.com/kiwiirc/kiwiirc/issues/294

**Comment by davidhedlund** [comment]
https://github.com/qwebirc/qwebirc/issues/345

**Comment by McInkay** [comment]
Why are you referencing other unrelated projects on this topic and just creating unnecessary email noise?

**Comment by dgw** [comment]
@McInkay Annoy enough people and they'll do what you want, right? :roll_eyes:

**Comment by bill-auger** [comment]
i can say without doubt that the librejs developers are not the
ones who evangelize it - david works on the FSF free software
directory, a distinct team from GNU librejs - i can not speakfor
him; but im sure that he did not intend to pester

this particular issue is a rare occasion when someone
representing GNU or the FSF has requested someone to seek
cooperation from an outside project; and i am probably the only
such person that has ever contacted the kiwi-irc team for any
reason in the past - i assume that david was merely following-up
on the issue, though perhaps too intrusively

i was asked simply to find out if librejs patches would be
accepted, because the FSF was interested in using this tool; and
i was fully satisfied with the original timely response - i also
agree that the cross-linking above was unnecessary and
inappropriate

it is almost always random users who take it upon themselves to
pester project maintainers publicly about licensing issues - i
often see vague or accusative licensing issues on other projects
posted by zealous users, with the result of annoying the devs -
and it is true that those people often do not offer to help
implement their feature requests in any way, as the most zealous
users are generally not programmers, from my experience - this
is unfortunate; but can not be avoided on a public access bug
tracker

**Comment by McInkay** [comment]
I am not trying to blame librejs for the actions of someone else, I just don't want unnecessary linking to other projects. As suggested, we (in theory) have nothing against some way to clarify things, but not in a way that increases bundle size dramatically or creates any other awkwardness, or even tries to change our license (we couldn't use non-MIT without some significant changes to dependencies).

So, with that in mind, our original point still stands, and I would appreciate that until something actually changes, unnecessary comments weren't left. If that happens, I'll be forced to lock the issue. We welcome future PRs. I have marked all comments after the original discussion, up till this one as off topic.

**Comment by bill-auger** [comment]
i apologize for my reply - it was mostly directed at the kiwi-irc thread - all these came into my email at once and i assumed they were all from the same thread and i replied to dgw

the kiwi-irc did explicitly state that the librejs team gives itself a bad reputatation by pestering people, and i am certain that is not true - the remark from dgw seemed to be echo-ing the same accusation though not as explicitly - i just wanted to clear that up - there was zero nagging happening here @dgw

**Comment by davidhedlund** [comment]
I'm sorry for the cross-linking and I will avoid this in the future.
